profile/src/main/java/com/scottlogic/deg/profile/reader/BaseConstraintReaderMap.java

```diff
@@ -0,0 +1,70 @@
+package com.scottlogic.deg.profile.reader;
+
+import java.util.*;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+public class BaseConstraintReaderMap implements ConstraintReaderMap {
+    private final Map<String, Map<String, ConstraintReader>> operatorAndValueToReadermap =
+        new HashMap<>();
+
+    public BaseConstraintReaderMap(Stream<ConstraintReaderMapEntrySource> providersToLoad) {
+        providersToLoad.forEach(p -> add(p.getConstraintReaderMapEntries()));
+    }
+
+    @Override
+    public ConstraintReader getReader(String operatorCode, String valueCode) {
+        if (operatorAndValueToReadermap == null || operatorAndValueToReadermap.isEmpty()) {
+            return null;
+        }
+        Map<String, ConstraintReader> valueToReaderMap = operatorAndValueToReadermap.get(operatorCode);
+        if (valueToReaderMap == null || valueToReaderMap.isEmpty()) {
+            throw new InvalidProfileException(
+                "Profile is invalid: no constraints known for " +
+                    operatorCode
+            );
+        }
+
+        // handle "is": "X", "value": "Y" cases
+        if (valueCode != null) {
+            ConstraintReader func = valueToReaderMap.get(valueCode);
+            if (func != null) {
+                return func;
+            }
+            for (Map.Entry<String, ConstraintReader> entry : valueToReaderMap.entrySet()) {
```

Comment:
Potential bug: `entrySet` returns a set, which does not contain an ordering. Therefore, on different runs, different entries from the map could be evaluated first, and hence return different entries from the map.

nb: In my experience almost every time Map.Entry<> is used explicitly there is almost always an issue hiding behind it.

Reply:
Fair point; it's impossible for there to be multiple exact matches, but there could potentially be multiple regex matches.  However I think at present it is acceptable to say "if the `ConstraintReaderMap` setup leads to profile syntax becoming ambiguous, behaviour is undefined."

In future when we formally open ourselves up to syntax extensibility, we would need to choose a proper behavioural strategy for determining which profile reading would be chosen if there are multiple matches.